[Woo POS] Coupons: Keep products and coupons scroll state separate

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes WOOMOB-322

## Description

This PR keeps scrolling state separate for Products vs Coupons:

https://github.com/user-attachments/assets/d4829a34-51f7-48d3-b5c9-172cf854a9f0

I've tried different approaches but most of them didn't work very well, I found that binding infinite scroll view and item list, as well as setting unique id to each element has worked best so far (and to be cleaner) but happy to iterate if you think of a better approach.

I wrap the scroll movement into an animation on purpose, as the UI feels pretty wonky if we do so, direct render feels best imho.

## Testing
- Load POS, scroll in Products
- Switch to Coupons, observe that we're at the top of the view, scroll around
- Switch back to Products, observe the previous scrolling state is saved. Same for back and forth.
- Observe that pagination and/or pull-to-refresh doesn't break behaviour.

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Bind item type and scroll position back to parent

iamgabrielma committed:
Wrap infinitescroll in proxy, update proxy position to last item via id

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Linting. use constants

staskus commented on WooCommerce/Classes/POS/Presentation/Item Selector/ItemList.swift:
```diff
@@ -12,6 +17,8 @@ struct ItemList<HeaderView: View>: View {
 
     // Navigation only uses this on iOS 17
     @State private var activeNavigationItem: POSItem? = nil
+    @State private var scrollPositions: [ScrollPositionKey: CGFloat] = [:]
```
It would be great if we could reuse `ItemListType` `enum`. We have so many of these similar enums already, and we already keep `selectedItemListType` state. 

```suggestion
    @State private var scrollPositions: [ScrollPositionKey: CGFloat] = [:]
```

iamgabrielma replied:
Good call, updated: https://github.com/woocommerce/woocommerce-ios/pull/15536/commits/534b34a9d5a79b033b3fb4f18120ce6bfce93505 . I had to make the object conform to Hashable to make it work, but no need for declare conformances as the associated values already are hashable as well 👍

staskus replied:
Nice! Thanks.

iamgabrielma committed:
Handle case when we load coupons for first time

The first time we switched to coupons it was moving also the state, as we only checked for saved positions. We assure it starts at the top the first tim we switch by passing a zero

iamgabrielma committed:
Reuse ItemListType

staskus committed:
Merge branch 'trunk' into task/woomob-322-separate-item-type-scroll-state

staskus commented on WooCommerce/Classes/POS/Presentation/Item Selector/ItemList.swift:
```diff
@@ -22,6 +24,15 @@ struct ItemList<HeaderView: View>: View {
 
+    private var scrollPositionKey: ItemListType {
+        switch itemsController {
+        case is PointOfSaleItemsController:
```
This is not a trustworthy check, since controller classes can change and we can miss this. I think better to pass `selectedItemListType` to `ItemList`.

staskus commented on WooCommerce/Classes/POS/Presentation/Item Selector/ItemList.swift:
```diff
@@ -22,6 +24,15 @@ struct ItemList<HeaderView: View>: View {
+    private var scrollPositionKey: ItemListType {
+        switch itemsController {
+        case is PointOfSaleItemsController:
+            return .products()
```
Also test with `searchProductsInPOS`. I don't think we need to save scroll position for search.

Right now, we have the same scroll position for both the product and the product search lists. It means if you scroll down in the product search and come back to the products list, the same scroll position will be maintained.

iamgabrielma replied:
> Right now, we have the same scroll position for both the product and the product search lists. It means if you scroll down in the product search and come back to the products list, the same scroll position will be maintained.

Good point, I'll check with both flags turned on. I agree we do not need to save scroll position for search as well.

staskus reviewed (commented):
Thank you!

I was really excited about this solution but unfortunately I see the list jumping from one position to another while switching, so it's not entirely precise. 
https://github.com/user-attachments/assets/31dc50e5-39ba-4ffa-b7b0-dd2bd2741e67

I recognize there are no straightforward solutions 🤔  One idea that popped was to use `TabView` as a container for `listView` within `ItemListView`. This way `SwiftUI` mechanisms would maintain the state of all these views. I'll experiment some more myself. 

```swift
  TabView(selection: $selectedItemListType) {
      listView(items).tag(ItemListType.products(search: false))
      listView(items).tag(ItemListType.products(search: true))
      listView(items).tag(ItemListType.coupons)
  }
  .tabViewStyle(.page(indexDisplayMode: .never))
  .animation(.none, value: selectedItemListType)
```

staskus commented:
I played around and made a draft PR https://github.com/woocommerce/woocommerce-ios/pull/15540/files, still requires testing. You could check it out 🤔 I think it's more reliable to use a native mechanism rather than count scroll position. However, it's not perfect and it doesn't maintain scroll position when switching between checkout and back, since we remove ItemListView when switching between .building and .finalizing stages.

@iamgabrielma let me know how you think

iamgabrielma commented:
> I played around and made a draft PR https://github.com/woocommerce/woocommerce-ios/pull/15540/files, still requires testing. You could check it out 🤔 I think it's more reliable to use a native mechanism rather than count scroll position. However, it's not perfect and it doesn't maintain scroll position when switching between checkout and back, since we remove ItemListView when switching between .building and .finalizing stages.

I really like your solution tbh, and we should move ahead with it:
- Native, we do not have to figure out scroll positions, save them, updating them, etc, ... 
- Less changes, and less complex changes
- The content is extracted to a ViewBuilder, which makes sense in the list view, we just keep the separation based on tags, which comes from selectedItemListType so it's natural

I'd vote for close this one in favor of #15540 , I don't think losing the scroll position when going back from checkout it's a problem at all, since the expectation is that we're done with building the order. Much preferred to keep it simple than over-engineering it for no reason really just to cover that case.

iamgabrielma commented:
[Closing in favor of 15540](https://github.com/woocommerce/woocommerce-ios/pull/15536#issuecomment-2822865166)!
